Slice [3/5] Add in slice reconciler and metrics

echiugoog opened the pull request:
- Add in slice metrics
  - add in slice attributes to nodepool metrics from node labels
  - add in slice attributes to jobset attributes  
  - add in slice "state" attribute from Reason field of Slice Ready condition
- Enable parallel testing of integration tests 
- Enhance GKE client mock to mock nodepools based on nodes observed#

echiugoog committed:
Use ginkgo explicitly in integration test, to enable control over parallel testing to speedup testing

echiugoog committed:
Copy in slice API from commit ca3afd0

echiugoog committed:
Simulation mode for megamon, mocks GKE and GCS calls

echiugoog committed:
Add in slice reconciler and metrics

echiugoog committed:
Add in slice reconciler and metrics

echiugoog committed:
correct PutRecords method to mock GCS

echiugoog committed:
Allow for optional controller name suffix
This allows for running multiple controllers in testenv

echiugoog committed:
Enable parallel testing of integration tests

echiugoog committed:
Enable parallel testing of integration tests

echiugoog committed:
integration tests for slice objects

echiugoog committed:
integration tests for slice objects

nstogner commented on internal/aggregator/aggregator.go:
```diff
@@ -139,6 +142,43 @@ func (a *Aggregator) Aggregate(ctx context.Context) error {
+		for _, s := range sliceList.Items {
+			attrs := records.Attrs{
+				SliceName:      s.Name,
+				SliceOwner:     s.Labels["tpu-provisioner.cloud.google.com/owner-name"],
```
Panic would happen if .Labels is nil

echiugoog replied:
Added nil check in f1693c9

nstogner commented on internal/controller/slice_reconciler.go:
```diff
@@ -0,0 +1,51 @@
+)
+
+// SliceReconciler reconciles a slice object
+type SliceReconciler struct {
```
What does this do?

echiugoog replied:
Nothing for now, just a placeholder and at high log levels add some verbose logging

nstogner replied:
Lets avoid adding in the code if it doesnt do anything

echiugoog replied:
Removed in 75099017237ddf3304386814e398856dc2c2d8f8

nstogner commented on internal/manager/manager.go:
```diff
@@ -77,12 +79,16 @@ type Config struct {
 	ProbeAddr            string
 	SecureMetrics        bool
 	EnableHTTP2          bool
+	EnableSimulation     bool
```
Can you explain how enabling simulation fits into the user workflow of the existing simulation environment?

echiugoog replied:
This will help with running megamon in an environment where GKE and GCE APIs are not available, such as Kind. More work is likely needed to work in a simulated environment.

echiugoog replied:
I updated the README in PR#21 and commit da5fdc886d73cb57af4abe114b84a01bc1e6b304 to make it clear that it is incomplete

nstogner commented on internal/metrics/metrics.go:
```diff
@@ -154,6 +157,15 @@ func OTELAttrs(attrs records.Attrs) []attribute.KeyValue {
+		otelAttrs = append(otelAttrs, attribute.String("slice.name", attrs.SliceName))
+	}
+	if attrs.SliceOwner != "" {
+		otelAttrs = append(otelAttrs, attribute.String("slice.owner", attrs.SliceOwner))
```
should this be `slice.owner.name`?

echiugoog replied:
I think that makes sense (matches the label better). Updated both the struct and metric in f6a2e49

nstogner commented on copied-slice-api/v1alpha1/groupversion_info.go:
```diff
@@ -0,0 +1,36 @@
+/*
```
Please add a readme or something indicating where these files came from. Also it would be good to name the directory something to indicate that these are actually copied in.

echiugoog replied:
Added in f1693c9 and renamed directory

echiugoog committed:
Extract slice name from node

echiugoog committed:
Add nil check on slice label lookup and make clear slice API is a copy

Add in brief README.md on origins of copied slice API

echiugoog committed:
Add nil check on slice label lookup and make clear slice API is a copy

Add in brief README.md on origins of copied slice API

echiugoog committed:
Refactor SliceOwner to SliceOwnerName, along with metric name

echiugoog committed:
use constant for tpu provisioner label names

echiugoog committed:
use constant for tpu provisioner label names

echiugoog committed:
Add in slice name and uid to jobset attributes

echiugoog committed:
Remove parallel integration testing target (for now)

echiugoog committed:
Allow for optional controller name suffix for slice reconciler

echiugoog committed:
Allow for optional controller name suffix for slice reconciler

echiugoog committed:
enhance GKE client mock to mock nodepools based on nodes observed

echiugoog committed:
Add slice attribute to nodepool metric from node label

echiugoog committed:
ignore not found error

echiugoog committed:
ignore not found error

echiugoog committed:
remove s390x and ppc64le target platforms from docker-buildx (#25)

echiugoog committed:
Slice [2/5] Simulation mode for megamon, mocks GKE and GCS calls (#21)

* Use ginkgo explicitly in integration test, to enable control over parallel testing to speedup testing
* Simulation mode for megamon, mocks GKE and GCS calls
* make clear EnableSimulation is incomplete for GKE

echiugoog committed:
update README.md on limitations of GCS and GKE mock client

echiugoog committed:
Merge branch 'main' of https://github.com/nstogner/megamon into slice-002

echiugoog committed:
update Dockerfile with copied-slice-api path

echiugoog committed:
README update on how to enable slice metrics

echiugoog committed:
add parallel integration target in Makefile

echiugoog committed:
fix attribute assignment

echiugoog committed:
Add in slice name and uid to jobset attributes

echiugoog committed:
remove slice_reconciler.go

echiugoog committed:
remove slice_reconciler.go

echiugoog committed:
Merge branch 'main' of https://github.com/nstogner/megamon into slice-002

echiugoog committed:
update reconcileEvents call after merging from main

echiugoog committed:
merge fix

echiugoog committed:
remove slice_reconciler.go

echiugoog committed:
merge fix

echiugoog committed:
Populate slice_state label in slice metrics from Reason field of Slice Ready condition.

Additionally simplify Slice Metrics Scenarios integration tests and
added a test for slice deletion

echiugoog committed:
Prevent jobset_nodes metrics from being emitted when "SliceEnabled" is true

echiugoog committed:
Cherry pick 9f5ac37, replace AfterSuite/AfterEach cleanup with Ginkgo's DeferCleanup to ensure reliable envtest environment teardown and prevent api-server and etcd process leaks

nstogner commented on config/rbac/role.yaml:
```diff
@@ -29,6 +29,20 @@ rules:
   - nodes/status
   verbs:
   - get
+- apiGroups:
```
This will break megamon deployed in non gsc clusters. See `default-slice/` in https://github.com/nstogner/tpu-provisioner/tree/7x-superslicing-release/config - we break out gsc and non gsc kustomizations

echiugoog replied:
Removed the new RBAC permissions here and elsewhere and created a "dev-slice" example for kustomize and Makefile target in 1acc898

nstogner replied:
"default" is what we install in clusters. I would rename this to `default-slice` b/c it will actually be used in prod clusters.

nstogner replied:
If you rename, dont forget to remove the dev config / SA

echiugoog replied:
I added in default-slice and updated README.md in b9de0d5
Verified by running kustomize:
```
❯ diff -u <(bin/kustomize build config/default) <(bin/kustomize build config/default-slice)
--- /proc/self/fd/16    2026-02-12 14:05:08.175867813 -0500
+++ /proc/self/fd/17    2026-02-12 14:05:08.175867813 -0500
@@ -101,6 +101,20 @@
   - jobsets/status
   verbs:
   - get
+- apiGroups:
+  - accelerator.gke.io
+  resources:
+  - slices
+  verbs:
+  - get
+  - list
+  - watch
+- apiGroups:
+  - accelerator.gke.io
+  resources:
+  - slices/status
+  verbs:
+  - get
```

nstogner commented on internal/aggregator/aggregator.go:
```diff
@@ -7,16 +7,21 @@ import (
 	jobset "sigs.k8s.io/jobset/api/jobset/v1alpha2"
 )
 
+// +kubebuilder:rbac:groups=accelerator.gke.io,resources=slices,verbs=get;list;watch
```
See other comment about rbac

echiugoog replied:
Will resolve in previous comment

nstogner commented on internal/manager/manager.go:
```diff
@@ -58,6 +59,7 @@ var (
 func init() {
 	//utilruntime.Must(clientgoscheme.AddToScheme(scheme))
 	utilruntime.Must(jobset.AddToScheme(scheme.Scheme))
+	utilruntime.Must(slice.AddToScheme(scheme.Scheme))
```
This will break megamon on non-gsc clusters

echiugoog replied:
This should work on clusters with and without slice CRD present, should be indirectly tested in https://github.com/nstogner/megamon/pull/24

nstogner replied:
How is it tested in #24? I see the CRD present in the integration tests.

echiugoog replied:
Added simple test in a5edabdda97b332b572b9fb8642af521127f3e7e to confirm init() will run w/o issue

nstogner replied:
I am failing to see how that works based on the docs/code for Must(). Can you provide a description of what is happening? https://pkg.go.dev/k8s.io/apimachinery/pkg/util/runtime#Must

nstogner commented on internal/metrics/metrics.go:
```diff
@@ -169,6 +172,15 @@ func OTELAttrs(attrs records.Attrs) []attribute.KeyValue {
+	if attrs.SliceName != "" {
+		otelAttrs = append(otelAttrs, attribute.String("slice.name", attrs.SliceName))
+	}
+	if attrs.SliceOwnerName != "" {
```
Are you missing the mapping of SliceOwnerNamespace which you also included in Attrs?

echiugoog replied:
Added now in fa7a44d

nstogner reviewed (changes requested):
Comments are mainly to avoid breaking changes. It would be good to lump in tests with the same PR as the functionality.

Also, how are you planning on addressing interruption metrics like MTTR? Those are implemented via Slice deletion and recreation. However, this will have the effect of removing the events from the datastore and each Slice recreation will be seen as a new event stream. See https://github.com/nstogner/megamon/blob/18f8450988843163c8ab61e9db8182938746fbbd/internal/records/events.go#L191C1-L196C3

```
	for key := range events {
		if _, ok := ups[key]; !ok {
			delete(events, key)
			changed = true
		}
	}
```

echiugoog committed:
Add in additional RBAC permissions for slice thru kustomize. Add a new target in Makefile to use example kustomize config 'dev-slice'

echiugoog commented:
Rebased the chain of PRs behind this onto slice-002 and force pushed.

echiugoog committed:
Add in slice_owner_namespace attribute to metrics

echiugoog committed:
remove redundant sliceUidMap

echiugoog committed:
Omit sliceUID from metric labels

nstogner commented on internal/gkeclient/client_mock.go:
```diff
@@ -2,22 +2,65 @@ package gkeclient
+	})
+
+	// Regex to match [nodepool_name]-n-n-n
+	re := regexp.MustCompile(`^(.*)-\d+-\d+-\d+(?:-\d+)*$`)
```
Use the nodepool name label instead

echiugoog replied:
Done in 1eb04aa

nstogner commented on internal/manager/manager.go:
```diff
@@ -58,11 +59,14 @@ var (
 type Config struct {
 	MetricsPrefix string
 
+	OptionalControllerSuffix string
```
What is this for?

echiugoog replied:
When running multiple controllers in testenv I think, the controller names needed to be unique.

nstogner replied:
I dont think that is the case, there is no api concept of a controller to enforce uniqueness in the k8s api. Lets remove this config.

nstogner commented on internal/metrics/metrics.go:
```diff
@@ -95,11 +95,20 @@ func Init(ctx context.Context, r Reporter, interval time.Duration, unknownThresh
+		jobsetNodeObservables, observeJobsetNodes = mustRegisterUpnessMetrics(Prefix+".jobset.nodes", meter, unknownThreshold)
+	}
 	nodePoolObservables, observeNodePools := mustRegisterUpnessMetrics(Prefix+".nodepool", meter, unknownThreshold)
+	sliceObservables, observeSlices := mustRegisterUpnessMetrics(Prefix+".slice", meter, unknownThreshold)
```
Shouldnt this be conditionally enabled using an if else connected to line 100? `if slicedEnabled {} else {`

echiugoog replied:
done in d3bbf5c

nstogner commented on internal/metrics/metrics.go:
```diff
@@ -95,11 +95,20 @@ func Init(ctx context.Context, r Reporter, interval time.Duration, unknownThresh
+		observables = append(observables, jobsetNodeObservables...)
+	}
 	observables = append(observables, nodePoolObservables...)
+	observables = append(observables, sliceObservables...)
```
Same comment as above

echiugoog replied:
Fixed in d3bbf5c

nstogner commented on internal/metrics/metrics.go:
```diff
@@ -118,8 +127,11 @@ func Init(ctx context.Context, r Reporter, interval time.Duration, unknownThresh
+			observeJobsetNodes(ctx, o, report.JobSetNodesUp, report.JobSetNodesUpSummaries)
+		}
 		observeNodePools(ctx, o, report.NodePoolsUp, report.NodePoolsUpSummaries)
+		observeSlices(ctx, o, report.SlicesUp, report.SlicesUpSummaries)
```
See same comment about conditional enablement

echiugoog replied:
Addressed in d3bbf5c, leaving this last comment open if we find more areas that need to be wrapped

nstogner commented on internal/metrics/metrics.go:
```diff
@@ -169,6 +181,21 @@ func OTELAttrs(attrs records.Attrs) []attribute.KeyValue {
+	if attrs.SliceOwnerKind != "" {
+		otelAttrs = append(otelAttrs, attribute.String("slice.owner.kind", attrs.SliceOwnerKind))
+	}
+	if attrs.SliceState != "" {
```
This will create separate metric streams whenever slice state changes. This is different from how all of the other metrics work: they all keep a singular metric stream that transitions states by transitioning values.

nstogner commented on internal/records/report.go:
```diff
@@ -34,6 +38,12 @@ type Attrs struct {
+
+	SliceName           string `json:"sliceName"`
+	SliceOwnerKind      string `json:"sliceOwnerKind"`
+	SliceState          string `json:"sliceState"`
```
See other comment about setting this metric label

nstogner commented on Makefile:
```diff
@@ -81,6 +81,12 @@ test-integration: $(INTEGRATION_TEST_DEPS) ## Run tests.
 test-integration-verbose: $(INTEGRATION_TEST_DEPS) ## Run tests.
 	$(INTEGRATION_TEST_CMD) -v --output-interceptor-mode=none ./test/integration
 
+test-integration-parallel: $(INTEGRATION_TEST_DEPS) ## Run tests.
```
How slow were the tests before? A lot of times introducing parallelism into tests can cause complications that are not worth it. Not to mention logs can get jumbled.

nstogner commented on Makefile:
```diff
@@ -152,10 +158,20 @@ deploy: manifests kustomize ## Deploy controller to the K8s cluster specified in
 	$(KUSTOMIZE) build config/dev | $(KUBECTL) apply -f -
 
+.PHONY: deploy-slice
+deploy-slice: manifests kustomize ## Deploy controller to the K8s cluster specified in ~/.kube/config.
```
`deploy-slice` should not deploy a dev kustomization with dev specific config

echiugoog replied:
"deploy" target (https://github.com/nstogner/megamon/blob/18f8450988843163c8ab61e9db8182938746fbbd/Makefile#L151-L153) uses config/dev, so "deploy-slice" is just a dev variant with slice support

nstogner replied:
lol, whoops that was a mistake too

nstogner replied:
Please rename both of these to *-dev if you want to keep the targets

echiugoog replied:
Fixed in 4dd8063, which also fixes missing LDFLAGS in docker build

nstogner commented on internal/controller/jobset_reconciler.go:
```diff
@@ -33,8 +34,12 @@ func (r *JobSetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 // SetupWithManager sets up the controller with the Manager.
 func (r *JobSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	name := "jobset"
+	if r.Name != "" {
```
what is this for?

echiugoog replied:
This was in case a name wasn't supplied, this shouldn't happen, we can remove that bit and update to something like:
```
	return ctrl.NewControllerManagedBy(mgr).
		For(&corev1.Pod{}).
		Named(r.name).
		Complete(r)
```

nstogner replied:
lgtm either way

nstogner commented on internal/controller/node_reconciler.go:
```diff
@@ -26,7 +27,12 @@ func (r *NodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.
 // SetupWithManager sets up the controller with the Manager.
 func (r *NodeReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	name := "node"
+	if r.Name != "" {
```
what is this for?

nstogner commented on internal/controller/pod_reconciler.go:
```diff
@@ -99,7 +100,12 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
 // SetupWithManager sets up the controller with the Manager.
 func (r *PodReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	name := "pod"
+	if r.Name != "" {
```
see same comment on other reconcilers

nstogner commented on internal/gkeclient/client_mock.go:
```diff
@@ -2,22 +2,65 @@ package gkeclient
 
 func (m *mockGKEClient) ListNodePools(ctx context.Context) ([]*containerv1beta1.NodePool, error) {
-	return m.nodePools, nil
+	// build fake GKE node pool object based on k8s nodes present, assumes a node naming convention
```
This looks pretty useful!

echiugoog committed:
manager init test

echiugoog committed:
add in default-slice configs for kustomize

echiugoog committed:
use node label for nodepool name vs fragile regex in mock GKE client

echiugoog committed:
Add in more conditionals around SliceEnabled

nstogner commented on config/default-slice/config.yaml:
```diff
@@ -0,0 +1,20 @@
+data:
+  config.json: |
+    {
+      "MetricsPrefix": "megamon.slicedev",
```
Customer will be deploying off of this kustomization dir. This will bifurcate their metrics and prefix everything with "slicedev"

nstogner replied:
This file does not look used, can we remove?

echiugoog replied:
Removed the entire dir in 61009a1

nstogner commented on config/dev-slice/kustomization.yaml:
```diff
@@ -0,0 +1,38 @@
+resources:
+- ../default
```
This should build from `../default-slice` and not duplicate the patches that are done in there (see below in this file)

echiugoog replied:
"default-slice" is removed in 61009a1, so there should be no duplicate patches now

nstogner replied:
I actually think we need it, thats how we will deploy this

echiugoog replied:
Added back to d64e575

Diff using kustomize shows the added rbac patch is working and no other changes:
```
❯ diff -u <(bin/kustomize build config/default) <(bin/kustomize build config/default-slice)
--- /proc/self/fd/16    2026-02-19 16:55:26.629397710 -0500
+++ /proc/self/fd/17    2026-02-19 16:55:26.629397710 -0500
@@ -382,6 +382,20 @@
   - jobsets/status
   verbs:
   - get
+- apiGroups:
+  - accelerator.gke.io
+  resources:
+  - slices
+  verbs:
+  - get
+  - list
+  - watch
+- apiGroups:
+  - accelerator.gke.io
+  resources:
+  - slices/status
+  verbs:
+  - get
 ---
 apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRole
```

nstogner commented on test/integration/cases_test.go:
```diff
@@ -328,32 +356,56 @@ var _ = Describe("Nodepool metrics", func() {
 
-	})
-})
+		It("should update slice_name in nodepool metrics when a node has the slice label", func() {
```
Are we sure we want to do this? This will break apart the nodepool metrics stream into N streams, one per slice_name over time to infinity as that nodepool is resliced

echiugoog replied:
Good catch.  Removed in 4535fa4

nstogner commented on test/integration/cases_test.go:
```diff
@@ -486,13 +540,81 @@ var _ = Describe("JobSet metrics", func() {
+			},
+		}
+
+		It("should update the JobSet metrics with the slice name", func() {
```
Same question as before for nodepools with relabelling based on slice, but in addition: JobSet to Slice is 1:N. How is that accounted for here? It seems to assert on a 1:1 relationship.

echiugoog replied:
Removed in 4535fa4

nstogner commented on test/integration/cases_test.go:
```diff
@@ -632,8 +765,262 @@ func expectedMetricsForJobSet(js *jobset.JobSet, tpuTopology string) upnessMetri
+			}
+
+			By("updating the slice status to interrupted")
+			updateSliceStatus(s, "SliceInterrupted", metav1.ConditionFalse)
```
Does this status match reality?

nstogner replied:
When does a Slice go into "SliceInterrupted" ?

echiugoog replied:
The reason/state is passthru. Interruption happens when READY condition goes false
Reworded a bit in 5f792fc

echiugoog replied:
Removed slice state in 2cdd8b3 so we don't emit multiple series

nstogner commented on test/integration/cases_test.go:
```diff
@@ -632,8 +765,262 @@ func expectedMetricsForJobSet(js *jobset.JobSet, tpuTopology string) upnessMetri
+			}
+
+			By("updating the slice status to recovered")
+			updateSliceStatus(s, "SliceRecovered", metav1.ConditionTrue)
```
Can Slices go into Interrupted states and then Recovered states?

echiugoog replied:
It's not expected to happen and shouldn't be possible. Removed in 903a08e

nstogner commented on test/integration/cases_test.go:
```diff
@@ -632,8 +765,262 @@ func expectedMetricsForJobSet(js *jobset.JobSet, tpuTopology string) upnessMetri
+	})
+})
+
+func expectedMetricsForSlice(s *slice.Slice) upnessMetrics {
```
Remove this func

echiugoog replied:
Removed usage in 5f792fc

echiugoog committed:
add ldflags to Dockerfile when building megamon

echiugoog committed:
update deploy targets with dev suffix, add in LDFLAGS for docker builds

echiugoog committed:
remove default-slice config

echiugoog committed:
Remove slice label from jobset metric
Remove slice_name from nodepool metric

echiugoog committed:
update Slice State Transition Scenarios test to use known reason values

echiugoog committed:
remove slice recovery test, as failed slices are not recoverable

echiugoog committed:
add back in default-slice

nstogner commented:
LGTM, thanks!
